text/0762-invokeHelper-property-thunks.md

```diff
@@ -0,0 +1,176 @@
+---
+Stage: Accepted
+Start Date: 2021-08-11
+Release Date: Unreleased
+Release Versions:
+  ember-source: vX.Y.Z
+  ember-data: vX.Y.Z
+Relevant Team(s): Ember.js
+RFC PR: https://github.com/emberjs/rfcs/pull/762
+---
+
+<!---
+Directions for above:
+
+Stage: Leave as is
+Start Date: Fill in with today's date, YYYY-MM-DD
+Release Date: Leave as is
+Release Versions: Leave as is
+Relevant Team(s): Fill this in with the [team(s)](README.md#relevant-teams) to which this RFC applies
+RFC PR: Fill this in with the URL for the Proposal RFC PR
+-->
+
+# Add argument-based thunks to invokeHelper
+
+## Summary
+
+`invokeHelper` (from `@ember/helper` (re-export from `@glimmer/runtime`), introduced in [RFC 626](https://github.com/emberjs/rfcs/pull/626))
+is a low-level utility for helping library-authors create reactive wrapping abstractions, such as _Resources_.
+
+However, `invokeHelper`'s third argument, the thunk, entangles all tracked data during creation
+of the helper and does not allow lazy-entanglement upon access. This RFC proposes a solution
+to enable lazy-entanglement of tracked data when using `invokeHelper`.
+
+```js
+import Component from '@glimmer/component';
+import { cached } from '@glimmer/tracking';
+import { invokeHelper } from '@ember/helper';
+import { getValue } from '@glimmer/tracking/primitives/cache';
+
+export default class MyComponent extends Component {
+  @tracked op = '+';
+  @tracked left = 1;
+  @tracked right = 2;
+
+  calculator = invokeHelper(this, Calculator, {
+    named: {
+      operation: () => this.op;
+    },
+    positional: [
+      () => this.left,
+      () => this.right,
+    ],
+  });
+
+  get calculatorValue() {
+    return getValue(this.calculator);
+  }
+}
+
+// Assume there is a Helper manager registered that knows what to do with this
+// and getValue(this.calculator) returns an instance of this class
+// (instead of calling compute like on the default Helper class)
```

Comment:
I have a reproduction of this in a stackblitz, but, you'd need to do some work to get it running :cry: 
https://github.com/stackblitz/webcontainer-core/issues/294